state-chain/ethereum-eip712/src/lib.rs

```diff
@@ -0,0 +1,409 @@
+							// appended by its index in the tuple
+							name: type_name.clone() + "__" + &i.to_string(),
+							r#type: type_name,
+						})
+					})
+					.collect::<Result<Vec<_>, _>>()?;
+				(
+					// In case of tuple, we decide to name it "UnnamedTuple_{first 4 bytes of hash
+					// of the fields}" since tuples, although supported in solidity, cant be
+					// easily displayed in metamask. Naming it so will display it in metamask
+					// which will indicate to the signer that this is indeed a tuple. The 4
+					// bytes of hash is just to avoid name collisions in case there are
+					// multiple unnamed tuples.
+					"UnnamedTuple__".to_string() +
+						&hex::encode(&keccak256(format!("{fields:?}"))[..4]),
+					fields,
+				)
+			},
+			(TypeDef::Primitive(type_def_primitive), ValueDef::Primitive(_p)) => (
+				match type_def_primitive {
+					TypeDefPrimitive::Bool => "bool".to_string(),
+					TypeDefPrimitive::Char => "string".to_string(),
+					TypeDefPrimitive::Str => "string".to_string(),
+					TypeDefPrimitive::U8 => "uint8".to_string(),
+					TypeDefPrimitive::U16 => "uint16".to_string(),
+					TypeDefPrimitive::U32 => "uint32".to_string(),
+					TypeDefPrimitive::U64 => "uint64".to_string(),
+					TypeDefPrimitive::U128 => "uint128".to_string(),
+					TypeDefPrimitive::U256 => "uint256".to_string(),
+					TypeDefPrimitive::I8 => "int8".to_string(),
+					TypeDefPrimitive::I16 => "int16".to_string(),
+					TypeDefPrimitive::I32 => "int32".to_string(),
+					TypeDefPrimitive::I64 => "int64".to_string(),
+					TypeDefPrimitive::I128 => "int128".to_string(),
+					TypeDefPrimitive::I256 => "int256".to_string(),
+				},
+				vec![],
+			),
+			(TypeDef::Compact(type_def_compact), _) => (
+				recursively_construct_types(v.clone(), type_def_compact.type_param, types)?,
+				vec![],
+			),
+			// this is only used when scale-info's bitvec feature is enabled and since we dont use
+			// that feature, this variant should be unreachable.
+			(TypeDef::BitSequence(_), _) => return Err("Unreachable"),
+
+			_ => return Err("Type and Value do not match"),
+		};
+
+	// If there are generic parameters to this type, append uniqueness to the type name to avoid
+	// collisions
+	if t.type_params.len() > 0 {
+		type_name = type_name + "__" + &hex::encode(&keccak256(format!("{fields:?}"))[..4]);
+	}
+
+	//TODO: maybe use the full path as the type name to avoid collisions due to same name types in
+	// different paths
+
+	// Only insert if there are fields (to avoid empty struct definitions)
+	if !fields.is_empty() {
+		types.insert(type_name.clone(), fields);
+	}
+
+	Ok(type_name)
+}
+
+fn process_composite<C: Clone>(
+	fields: Vec<Field>,
+	comp_value: Composite<C>,
+	types: &mut BTreeMap<String, Vec<Eip712DomainType>>,
+) -> Result<Vec<Eip712DomainType>, &'static str> {
+	match comp_value {
+		Composite::Named(fs) => {
+			let fs_map = fs.into_iter().collect::<BTreeMap<_, _>>();
+			fields
+				.clone()
+				.into_iter()
```

Comment:
Seems like a lot of unnecessary cloning... but also, why do we collect into a map? Do we not want to preserve the order of the fields?

*edit: I see it's so that we can look up the types by field name, but elsewhere we assume the ordering is preserved...*